### PR TITLE
[JIT Kernel]Support fused_add_rmsnorm in JIT Kernel

### PR DESCRIPTION
**Pull request by HydraQYH**
## Motivation
There are two differences from the Flashinfer implementation:

1. Avoid storing/loading inp+res in shared memory
2. Using 256-bit LDG on B200

## Modifications

- python/sglang/jit_kernel/norm.py
- python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh
- python/sglang/jit_kernel/tests/test_fused_add_rmsnorm.py
- python/sglang/jit_kernel/include/sgl_kernel/vec.cuh

## Accuracy Tests
H20(with CUDA 12.9):
image
B200(witch CUDA 13.1):
image

## Benchmarking and Profiling
H20(with CUDA 12.9):
[h20_fused_add_rmsnorm.txt](https://github.com/user-attachments/files/24859003/h20_fused_add_rmsnorm.txt)
B200(witch CUDA 13.1):
[b200_fused_add_rmsnorm.txt](https://github.com/user-attachments/files/24859009/b200_fused_add_rmsnorm.txt)

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
3. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
4. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
5. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+#include <cooperative_groups/reduce.h>
+#include <tvm/ffi/container/tensor.h>
+
+#include <cooperative_groups.h>
+#include <type_traits>
+
+namespace {
+
+union U16B_bf162 {
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh]: You may try `device::AlignedVector<bf16x2_t, 4>`. Currently, we only allow at most 16B aligned vector, but it's ok to permit 32B aligned vector (e.g. `device::AlignedVector<bf16x2_t, 8>`) if it's supported on certain devices (like B200).

File: python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+#else
+  using V = longlong4;
+#endif
+};
+
+template <typename T>
+SGL_DEVICE T rms(T& val, T& weight, float rsqrt_square_sum);
+
+template <>
+SGL_DEVICE __nv_bfloat162 rms<__nv_bfloat162>(__nv_bfloat162& val, __nv_bfloat162& weight, float rsqrt_square_sum) {
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh]: You may take a look at `<sgl_kernel/type.cuh>`, where we define many interesting type traits and support type casting. Try `device::cast` function, it's just like static_cast.

File: python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        float2 val = __bfloat1622float2(u.compute_unit[i]);
+        float2 res = __bfloat1622float2(u_res.compute_unit[i]);
+        float2 inp_res = make_float2(val.x + res.x, val.y + res.y);
+        acc_square.x += inp_res.x * inp_res.x;
+        acc_square.y += inp_res.y * inp_res.y;
+        u.compute_unit[i] = __float22bfloat162_rn(inp_res);
+      }
+    } else if constexpr (std::is_same_v<T, __half>) {
+      for (int i = 0; i < inner_loop; i++) {
+        float2 val = __half22float2(u.compute_unit[i]);
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh]: Similarly, for type casting, try to use `cast` function to unify the logic here.

File: python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh (hunk 1 above)
Comment (HydraQYH) [reply python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh]: Done.

File: python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh (hunk 3 above)
Comment (HydraQYH) [reply python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh]: Done.

File: python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh (hunk 2 above)
Comment (HydraQYH) [reply python/sglang/jit_kernel/csrc/elementwise/fused_add_rmsnorm.cuh]: Done.

File: python/sglang/jit_kernel/tests/test_fused_add_rmsnorm.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+def flashinfer_fused_add_rmsnorm(
+    input: torch.Tensor, residual: torch.Tensor, weight: torch.Tensor, eps: float
+) -> None:
+    from flashinfer.norm import fused_add_rmsnorm
+
+    fused_add_rmsnorm(input, residual, weight, eps=eps)
+
+
+BS_LIST = [2**n for n in range(0, 14)]
+BS_LIST += [x + 1 + i for i, x in enumerate(BS_LIST)]
```
[the remark this answers is not in the document]
Comment (HydraQYH) [reply python/sglang/jit_kernel/tests/test_fused_add_rmsnorm.py:24]: Copy from python/sglang/jit_kernel/tests/test_rmsnorm.py.

File: python/sglang/jit_kernel/tests/test_fused_add_rmsnorm.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+HIDDEN_SIZE_LIST = [512, 1024, 1536, 2048, 3072, 4096, 5120, 6144, 7168, 8192]
+DEVICE = "cuda"
+DTYPE = torch.bfloat16
+
+
+@pytest.mark.parametrize(
+    "batch_size,hidden_size", list(itertools.product(BS_LIST, HIDDEN_SIZE_LIST))
+)
+def test_fused_add_rmsnorm(batch_size: int, hidden_size: int) -> None:
```
[the remark this answers is not in the document]
Comment (HydraQYH) [reply python/sglang/jit_kernel/tests/test_fused_add_rmsnorm.py:33]: Copy from python/sglang/jit_kernel/tests/test_rmsnorm.py.

File: python/sglang/jit_kernel/benchmark/bench_fused_add_rmsnorm.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import os
+
+import torch
+import triton
+import triton.testing
+from flashinfer import fused_add_rmsnorm as fi_fused_add_rmsnorm
+
+from sglang.jit_kernel.norm import fused_add_rmsnorm as jit_fused_add_rmsnorm
+
+IS_CI = (
```
Comment (BBuf) [inline python/sglang/jit_kernel/benchmark/bench_fused_add_rmsnorm.py]: IS_CI has been refactored in https://github.com/sgl-project/sglang/pull/17118

**Review by BBuf — APPROVED** [review]

File: python/sglang/jit_kernel/benchmark/bench_fused_add_rmsnorm.py (hunk 6 above)
Comment (HydraQYH) [reply python/sglang/jit_kernel/benchmark/bench_fused_add_rmsnorm.py]: Done.

File: python/sglang/jit_kernel/include/sgl_kernel/vec.cuh [hunk 7]
```diff
@@ -45,7 +45,7 @@ template <typename T, std::size_t N>
 struct AlignedVector {
  private:
   /// NOTE: 1. must be pow of two 2. 16 * 8 = 128 byte, which is the max vector size supported by most devices
-  static_assert((N > 0 && (N & (N - 1)) == 0) && sizeof(T) * N <= 16, "CUDA only support at most 128B vector op");
+  static_assert((N > 0 && (N & (N - 1)) == 0) && sizeof(T) * N <= 32, "CUDA only support at most 256B vector op");
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/include/sgl_kernel/vec.cuh:48]: This is actually very tricky. Before B200, the max load/store size is 128bit. On and after B200, the max size is 256bit. On H200, developers may misuse this API and lead to poor performance.

It's okay to keep this behavior. But in the future we need to forbid using 256bit load/store on H200.

**Review by DarkSharpness — APPROVED** [review]

File: python/sglang/jit_kernel/include/sgl_kernel/vec.cuh (hunk 7 above)
Comment (HydraQYH) [reply python/sglang/jit_kernel/include/sgl_kernel/vec.cuh:48]: Yeah. I think we could use `__CUDA_ARCH__` for conditional compilation. However, this macro only works on device code, which might not meet our requirements.
